### PR TITLE
Fix false positive with shodan search module

### LINKED ISSUE #13598: Shodan API key error with metasploit
## Steps to reproduce

How'd you do it?

1. ...  msfconsole search shodan use auxiliary/gather/shodan_search 
2. ... set SHODAN_APIKEY set QUERY run

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

## Expected behavior it should scan

What should happen? supposed to give results of scan

## Current behavior - gives error

What happens instead?  error message : Auxiliary aborted due to failure: bad-config: 401 Unauthorized. Your SHODAN_APIKEY is invalid

I HAVE A SHODAN PAID MEMBERSHIP

You might also want to check the last ~1k lines of
`/opt/metasploit/apps/pro/engine/config/logs/framework.log` or
`~/.msf4/logs/framework.log` for relevant stack traces

### Metasploit version

Get this with the `version` command in msfconsole (or `git log -1 --pretty=oneline` for a source install).   Processing version.rc for ERB directives.
resource (version.rc)> version
Framework: 5.0.91-dev
Console  : 5.0.91-dev

### I installed Metasploit with:
- [ x] Kali package via apt  YES
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [ x] Source install (please specify ruby version)    ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux-gnu]

### OS

What OS are you running Metasploit on?   Linux kali 5.6.0-kali2-amd64 #1 SMP Debian 5.6.14-1kali1 (2020-05-25) x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by adfoster-r7**
closes https://github.com/rapid7/metasploit-framework/issues/13598

## Before

The shodan search module would raise errors to the user when the response payload from shodan contains `<title>401 Unauthorized<\/title>` anywhere in the payload. This caused false positives as an arbitrary host could include this string within its http response.

## After

The module now correctly checks the server's 401 status code. The module has been converted to use the Rex API to allow for `set HttpTrace true` to be enabled. 

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] Use the shodan module:
```
use auxiliary/gather/shodan_search
set SHODAN_APIKEY YOUR_KEY_HERE
set QUERY hipcam
run
```
- [ ] Verify that the search comes back successfully

### CODE REVIEW

File: modules/auxiliary/gather/shodan_search.rb [hunk 1]
```diff
@@ -41,19 +42,27 @@ def initialize(info = {})
...
+      'uri' => '/shodan/host/search',
+      'SSL' => true,
+      'vars_get' => {
+        'query' => query,
+        'key' => apikey,
+        'page' => page.to_s
+      }
+    })
+
+    if res and res.code == 401
```
Comment (wvu-r7) [inline modules/auxiliary/gather/shodan_search.rb:65]:
```suggestion
    if res && res.code == 401
```

File: modules/auxiliary/gather/shodan_search.rb [hunk 2]
```diff
@@ -41,19 +42,27 @@ def initialize(info = {})
         OptRegexp.new('REGEX', [true, 'Regex search for a specific IP/City/Country/Hostname', '.*'])
 
       ])
+
+    deregister_http_client_options
```
Comment (wvu-r7) [inline modules/auxiliary/gather/shodan_search.rb:46]: Good call!

File: modules/auxiliary/gather/shodan_search.rb [hunk 3]
```diff
@@ -41,19 +42,27 @@ def initialize(info = {})
...
+      'method' => 'GET',
+      'rhost' => 'api.shodan.io',
+      'rport' => 443,
+      'uri' => '/shodan/host/search',
+      'SSL' => true,
+      'vars_get' => {
+        'query' => query,
+        'key' => apikey,
+        'page' => page.to_s
+      }
```
Comment (wvu-r7) [inline modules/auxiliary/gather/shodan_search.rb:62]: :+1:

**Review by wvu — APPROVED** [review]

**Comment by h00die** [comment]
+1 to @adfoster-r7 for working through this problem!

There is a second shodan module, you may want to take a look at that one too, not sure if it suffers the same problem

**Comment by wvu-r7** [comment]
Looks like `shodan_honeyscore` uses `Rex::Proto::Http::Client`, and it looks fine.

**Comment by wvu-r7** [comment]
@adfoster-r7: You are welcome to merge this. <3

**Comment by adfoster-r7** [comment]
## Release Notes

Fixes a false positive where the Shodan Search module would raise an unauthenticated issue if any search target had the substring of 401 Unauthorized within its html response.

**Pull request merged**